### PR TITLE
BC-7 # `bm bmp deploy`, with refactoring

### PR DESCRIPTION
**Pull request by jokeyrhyme**
- `bm bmp deploy`

- refactor so that data such as credentials or scopes are no longer passed all the way through the call graph from top to bottom

    - simplifies testing, and limits the concerns of each individual function

- environment variables to control a few test-related but useful behaviours

- changed `getOrigin()` so that it wouldn't swallow errors

- suggest reviewing individual commits, skip any that don't start with "BC-7 # ..."

### CODE REVIEW

**Comment by jokeyrhyme** [comment]
Just waiting for BC-6 (#3) to be merged so that this can be `git rebase`d.

File: lib/api.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  return auth.read()
+    .then((a) => httpUtils.sendRequest({
+      credential: a.credential,
+      url: `${a.origin}/_api/v1/${options.type}/${options.id}`
+    }));
+}
+
+function putResource (options) {
+  options = options || {};
+  assertTypeAndId(options);
```
Comment (simonmarklar) [inline lib/api.js]: you could be super DRY here by making a wrapper function eg

    var wrapper = fn =>  options => { 
       /* common checking code here */ 
       return fn(options);
    }

then when you do exports wrap the functions that need the assertion

File: lib/auth.js [hunk 2]
```diff
@@ -11,65 +11,116 @@ const mkdirp = pify(require('mkdirp'));
...
+    });
+}
+
+function read () {
+  let origin;
+  return scope.read()
+    .then((projectScope) => {
+      origin = urlUtils.getOrigin(urlUtils.toHTTPS(projectScope) || '');
+    })
+    .then(() => readAll())
```
Comment (simonmarklar) [inline lib/auth.js]: i think there's no need for a new function here, i think this should achieve the same thing
    
    .then(readAll)

**Comment by simonmarklar** [comment]
I have a couple of minor comments that are not deal breakers.

so LGTM :+1:

**Comment by jokeyrhyme** [comment]
@simonmarklar like this?

**Pull request merged**